### PR TITLE
feat: adds a new method for requesting exercise permissions

### PR DESCRIPTION
**Pull request by ugurakin1**
### Changes
1. Adds a new method for requesting route data of a particular exercise session record following the guide from [Android docs](https://developer.android.com/health-and-fitness/guides/health-connect/develop/exercise-routes).
2. Modifies `requestPermission` to accept an optional flag that indicates requested permissions should include `write-route` permission. Removed the unused `providerPackageName` arg from the method as well.
3. Updates the example app to include a button for inserting a random exercise session with routes and another button that requests the route of the given exercise id.

### Issues
#34

### CODE REVIEW

File: android/src/main/java/dev/matinzd/healthconnect/HealthConnectModule.kt [hunk 1]
```diff
@@ -34,12 +34,21 @@ class HealthConnectModule internal constructor(context: ReactApplicationContext)
   @ReactMethod
   override fun requestPermission(
     permissions: ReadableArray,
-    providerPackageName: String,
```
Comment (ugurakin1) [inline android/src/main/java/dev/matinzd/healthconnect/HealthConnectModule.kt:37]: Wasn't used so replaced it.

File: example/src/App.tsx [hunk 2]
```diff
@@ -174,24 +216,27 @@ export default function App() {
...
+        {
+          accessType: 'write',
+          recordType: 'ExerciseSession',
+        },
+        {
+          accessType: 'read',
+          recordType: 'ExerciseSession',
+        },
+      ],
+      true
```
Comment (ugurakin1) [inline example/src/App.tsx:238]: This is the only change here

**Comment by ugurakin1** [comment]
Should solve #86, #127, #128 @matinzd. Whipped up in a rush, happy to address comments.

File: src/index.tsx [hunk 3]
```diff
@@ -97,16 +97,20 @@ export function openHealthConnectDataManagement(
  */
 export function requestPermission(
   permissions: Permission[],
-  providerPackageName = DEFAULT_PROVIDER_PACKAGE_NAME
+  includeRoute?: boolean
```
Comment (ugurakin1) [inline src/index.tsx:100]: Leaving this `null` seems to cause issues atm:
> Exception in native call
com.facebook.react.bridge.NativeArgumentsParseException: Attempt to invoke virtual method 'boolean java.lang.Boolean.booleanValue()' on a null object reference (constructing arguments for HealthConnect.requestPermission at argument index 1)

File: android/src/main/java/dev/matinzd/healthconnect/HealthConnectManager.kt [hunk 4]
```diff
@@ -62,16 +64,34 @@ class HealthConnectManager(private val applicationContext: ReactApplicationConte
...
+    throwUnlessClientIsAvailable(promise) {
+      coroutineScope.launch {
+        val exerciseRoute = HealthConnectPermissionDelegate.launchExerciseRouteAccessRequestDialog(recordId)
+        if (exerciseRoute != null) {
+          promise.resolve(ReactExerciseSessionRecord.parseExerciseRoute(exerciseRoute))
+        }
+        else{
+          promise.rejectWithException(ExerciseRouteAccessDenied())
+        }
+      }
```
Comment (ugurakin1) [inline android/src/main/java/dev/matinzd/healthconnect/HealthConnectManager.kt:91]: I'm getting some flaky results with exercises, everything returns permissions denied as in exerciseRoute is null and as is there is no way to know for certain if that's because there is no data or something is wrong with my patch/permissions etc. Maybe we should fetch the exercise record here and switch on `ExerciseRouteResult` to determine whether to request permission or reject with "NoDataException". Could go hand in hand with [this](https://github.com/matinzd/react-native-health-connect/issues/86#issuecomment-2402741908).

**Comment by ugurakin1** [comment]
Will look into this further this weekend

File: src/index.tsx [hunk 5]
```diff
@@ -96,17 +97,20 @@ export function openHealthConnectDataManagement(
  * @returns granted permissions
  */
 export function requestPermission(
-  permissions: Permission[],
-  providerPackageName = DEFAULT_PROVIDER_PACKAGE_NAME
+  permissions: (Permission | WriteExerciseRoutePermission)[]
 ): Promise<Permission[]> {
-  return HealthConnect.requestPermission(permissions, providerPackageName);
+  return HealthConnect.requestPermission(permissions);
```
Comment (ugurakin1) [inline src/index.tsx:102]: Initially replaced this with another arg, then removed the arg in favor of a special permission obj (see `src/types/index.ts`). It wasn't doing anything so didn't add it again but can undo the changes for the sake of not breaking anyone's code.

File: android/src/main/java/dev/matinzd/healthconnect/HealthConnectManager.kt (hunk 4 above)
Comment (ugurakin1) [reply android/src/main/java/dev/matinzd/healthconnect/HealthConnectManager.kt:90]: Turns out Samsung Health never writes exercise data so this is fine. Everything in the example app works as expected but need to test using our app and see if I can read exercise sessions/routes inserted via the example app as expected from our app.

**Comment by ugurakin1** [comment]
@matinzd Any chance you could take a look at this? I patched our own app and syncing worked as expected. It also touches a few other things that we don't use necessarily so a review would be much appreciated 🙏

File: src/index.tsx (hunk 3 above)
Comment (ugurakin1) [reply src/index.tsx]: Opted for a special type of permission rather than an additional argument so this isn't relevant anymore.

**Comment by matinzd** [comment]
Hey @ugurakin1 

Can you please rebase your branch?

**Comment by ugurakin1** [comment]
Done @matinzd

**Comment by ugurakin1** [comment]
Hey @matinzd , any updates on this/eta? 

Also, did you want me to just update so there aren't any conflicts or did you actually want me to rebase instead?

**Comment by matinzd** [comment]
Hey!

I am gonna try to release it this weekend. It doesn't matter, I don't want to see any conflicts.

**Comment by ugurakin1** [comment]
Let me know if you need anything from me @matinzd, happy to land a hand 👍

File: android/src/main/java/dev/matinzd/healthconnect/records/ReactExerciseSessionRecord.kt
Comment (matinzd) [inline android/src/main/java/dev/matinzd/healthconnect/records/ReactExerciseSessionRecord.kt:1]: I feel that these changes can be a breaking change.

Comment (ugurakin1) [reply android/src/main/java/dev/matinzd/healthconnect/records/ReactExerciseSessionRecord.kt:1]: Yeah it's not easy to say they aren't, but the return type in TS should be backwards compatible. Anything we can do to test or modify so it's not? Or are we okay to introduce a breaking change at this moment?

Comment (ugurakin1) [reply android/src/main/java/dev/matinzd/healthconnect/records/ReactExerciseSessionRecord.kt:1]: See `src/types/base.types.ts`

Comment (matinzd) [reply android/src/main/java/dev/matinzd/healthconnect/records/ReactExerciseSessionRecord.kt:1]: Yeah I saw that now. Can you please update docs as well to include the information about how such cases should be handled?

Comment (ugurakin1) [reply android/src/main/java/dev/matinzd/healthconnect/records/ReactExerciseSessionRecord.kt:1]: Taking a look now 👍

Comment (ugurakin1) [reply android/src/main/java/dev/matinzd/healthconnect/records/ReactExerciseSessionRecord.kt:1]: @matinzd Updated the PR. Let me know if the docs look okay. Might be offline for the rest of the day but can take a further look tomorrow, just let me know what you need from me 👍

Comment (ugurakin1) [reply android/src/main/java/dev/matinzd/healthconnect/records/ReactExerciseSessionRecord.kt:1]: Any luck taking a look at this @matinzd ?

File: src/index.tsx (hunk 5 above)
Comment (matinzd) [reply src/index.tsx:103]: This needs to be fixed later by myself. By default most people use health connect to get the data, but it should be possible to customize this.

**Comment by ugurakin1** [comment]
Massive thanks @matinzd 🙏
